nitransforms/io.py

```diff
@@ -0,0 +1,232 @@
+        volgeom = klass()
+        sa = volgeom.structarr
+        lines = string.splitlines()
+        for key in ('valid', 'filename', 'volume', 'voxelsize',
+                    'xras', 'yras', 'zras', 'cras'):
+            label, valstring = lines.pop(0).split(' = ')
+            assert label.strip() == key
+
+            val = np.genfromtxt([valstring.encode()],
+                                dtype=klass.dtype[key])
+            sa[key] = val.reshape(sa[key].shape) if val.size else ''
+
+        return volgeom
+
+
+class LinearTransform(StringBasedStruct):
+    template_dtype = np.dtype([
+        ('mean', 'f4', (3, 1)),       # x0, y0, z0
+        ('sigma', 'f4'),
+        ('m_L', 'f4', (4, 4)),
+        ('m_dL', 'f4', (4, 4)),
+        ('m_last_dL', 'f4', (4, 4)),
+        ('src', VolumeGeometry),
+        ('dst', VolumeGeometry),
+        ('label', 'i4')])
+    dtype = template_dtype
+
+    def __getitem__(self, idx):
+        val = super(LinearTransform, self).__getitem__(idx)
+        if idx in ('src', 'dst'):
+            val = VolumeGeometry(val)
+        return val
+
+    def to_string(self):
+        sa = self.structarr
+        lines = [
+            'mean      = {:6.4f} {:6.4f} {:6.4f}'.format(
+                *sa['mean'].flatten()),
+            'sigma     = {:6.4f}'.format(float(sa['sigma'])),
+            '1 4 4',
+            ('{:18.15e} ' * 4).format(*sa['m_L'][0]),
+            ('{:18.15e} ' * 4).format(*sa['m_L'][1]),
+            ('{:18.15e} ' * 4).format(*sa['m_L'][2]),
+            ('{:18.15e} ' * 4).format(*sa['m_L'][3]),
+            'src volume info',
+            self['src'].to_string(),
+            'dst volume info',
+            self['dst'].to_string(),
+        ]
+        return '\n'.join(lines)
+
+    @classmethod
+    def from_string(klass, string):
+        lt = klass()
+        sa = lt.structarr
+        lines = string.splitlines()
+        for key in ('mean', 'sigma'):
+            label, valstring = lines.pop(0).split(' = ')
+            assert label.strip() == key
+
+            val = np.genfromtxt([valstring.encode()],
+                                dtype=klass.dtype[key])
+            sa[key] = val.reshape(sa[key].shape)
+        assert lines.pop(0) == '1 4 4'
+        val = np.genfromtxt([valstring.encode() for valstring in lines[:4]],
+                            dtype='f4')
+        sa['m_L'] = val
+        lines = lines[4:]
+        assert lines.pop(0) == 'src volume info'
+        sa['src'] = np.asanyarray(VolumeGeometry.from_string('\n'.join(lines[:8])))
+        lines = lines[8:]
+        assert lines.pop(0) == 'dst volume info'
+        sa['dst'] = np.asanyarray(VolumeGeometry.from_string('\n'.join(lines)))
+        return lt
+
+
+class LinearTransformArray(StringBasedStruct):
+    template_dtype = np.dtype([
+        ('type', 'i4'),
+        ('nxforms', 'i4'),
+        ('subject', 'U1024'),
+        ('fscale', 'f4')])
+    dtype = template_dtype
+    _xforms = None
```

Comment:
I would not allow this to contain VOX2VOX. Meaning, if ``transform_code`` is 0, then the transforms are decomposed and the RAS2RAS extracted. If that is not possible because moving and/or reference VOX2RAS are missing, then raise an error.

That said, I'd be fine with a `NotImplementedError` when transform code is 0.

Reply:
since this builds off @effigies implementation, we should rope him in here. 

I assumed the scope of his LTA implementation is greater than just nitransforms' use-case, so we may want to still support vox2vox as a valid matrix. however, totally agree we should catch that case within the transforms module, and coerce it into ras2ras.

Reply:
We should definitely permit reading/writing non-RAS2RAS, even if we only ever store RAS2RAS. I vaguely recall I might have intended to store the incoming transform, so that a load-save round trip would not change the contents, and only convert at need, but don't feel bound by this.

Reply:
would the `mean` / `sigma` change if we convert the matrix between transform types?

Reply:
Please note I'm not saying we should only support `LINEAR_RAS_TO_RAS`, I'm saying we should not write (just write) `LINEAR_VOX_TO_VOX`.

VOX2VOX is a legacy method that only makes sense in the context of the early development of image registration. Why (and who) anyone would like to write VOX2VOX? There's literally nothing VOX2VOX can do that cannot be done with RAS2RAS.

judging by https://github.com/freesurfer/freesurfer/blob/d5ff65ce78fee3ef296cc0b4027360ba6f9721f1/utils/transform.cpp#L823, I don't think `sigma` or `mean` should change with RAS2RAS.

Reply:
I would be more compelled by some demonstration of better numerical stability or precision of VOX2VOX, but I would actually guess that's also going to work in the opposite way.

Reply:
After checking further, it seems mean and sigma are not necessary for RAS2RAS at all (https://github.com/freesurfer/freesurfer/blob/b156d5aee6df2c7027ea45d3824813f8dcc536ef/lta_convert/lta_convert.cpp#L337).